packages/headless/src/api/search/search-api-client.test.ts

```diff
@@ -542,4 +543,47 @@ describe('search api client', () => {
       );
     });
   });
+
+  describe('assigning searchUid to results', () => {
+    let mockResponse: SearchResponseSuccess;
+
+    beforeEach(() => {
+      const searchUid = 'hello';
+      const mockResult = buildMockResult();
+      delete (mockResult.searchUid as Partial<Result>).searchUid;
+      mockResponse = buildMockSearchResponse({
+        searchUid,
+        results: [mockResult],
+      });
+    });
+
+    it('for search', async () => {
+      const res = await doMockPlatformResponseAndAssertSuccess(mockResponse);
+      expect(res.results[0].searchUid).toBe(res.searchUid);
+    });
+
+    it('for recommendations', async () => {
+      const res = await doMockPlatformResponseAndAssertSuccess(
+        mockResponse,
+        () =>
+          searchAPIClient.recommendations(
+            buildRecommendationRequest(createMockRecommendationState())
+          )
+      );
+      expect(res.results[0].searchUid).toBe(res.searchUid);
+    });
+
+    it('for recommendations', async () => {
+      const res = await doMockPlatformResponseAndAssertSuccess(
+        mockResponse,
+        () =>
+          searchAPIClient.productRecommendations(
+            buildProductRecommendationsRequest(
+              buildMockProductRecommendationsState()
+            )
+          )
+      );
+      expect(res.results[0].searchUid).toBe(res.searchUid);
+    });
+  });
```

Comment:
**General thoughts**

I understand the value of this change, and that it is convenient solution for the UI. There are some impacts however.

Type-safety is being managed at run-time using unit tests. If someone uses the `Result` interface with a new endpoint, but forgets to shim, there is no unit test to catch the error.

For headless to stay consistent, any endpoint supporting a "load more" UX would need to apply the shim.

Recommendations do not support "load more", but we still need to shim so that the `Result` interface is accurate.

Product recommendations are similar: no "load more" => no need for a search uid, but we need to shim to respect the interface. Moreover, with Product Recommendations, the shimmed `Result` [is mapped](https://github.com/coveo/ui-kit/blob/master/packages/headless/src/features/product-recommendations/product-recommendations-actions.ts#L161), so the search uid is not being exposed.

**Summary**

We shim three places so that the `Result` interface stays accurate, but only one needs the shim today. Future contributors will need to be aware of the shim since there is no type-safety. This can be mitigated somewhat using dedicated interfaces.

At the same time, I think the best place to solve UI problems is in the UI. The dev implementing the template knows best what requires a rerender. e.g.

- If there is no highlighting, the `uniqueId` is enough.
- If there are query highlights, then hashing the `Result` object is enough.
- If there are facet value highlights as in the case of Atomic, then it's more complex and need to be taken into account when creating the key. A fine-grained approach can avoid certain rerenders though, creating a better experience.

Reply:
I feel like comparing results & their contexts may happen very often given the popularity of JSX (and that even non-JSX applications may want to selectively re-render elements for performance).

I think we should either provide a `searchUid` in the result, therefor assuming any change in other parts of the search should trigger a re-render, or we should provide a helper with some options. E.g.: `getComparisonId(result, { includeQuery: true, includeFacets: true, ... })`

Reply:
Having the searchUid in the results offers the possibility of associating a result with it's query, that can have other utilities, such as the possible future search impression. We can also say that currently we wrong data to UA as the `searchQueryUid` since it will be the latest fetch and not the `searchQueryUid` where that result has actually been fetched